### PR TITLE
Changing the OpenSMT API

### PR DESCRIPTION
**Pull request by blishko**
This PR aims to solve #135 

This is solved by moving the code working with this library to the main .cc file which is part of the executable only.

In addition, some refactoring has been done to better define the relations between the main components: MainSolver now becomes responsible was maintaining all other components (SimpSMTSolver, Theory, THandler, ...), but it is created with a specific Theory.

### CODE REVIEW

**Comment by blishko** [comment]
@aehyvari please check the changes and let me know if you think this would be OK.
There is also the question if we should do something about `MainSolver::readFormulaFromFile` method and `Interpret::parse_only` field.

**Comment by blishko** [comment]
The original PR has been split to two parts. The first part that dealt with restricting the dependency on Readline/Editline library has been moved to #151.

The aim of this second part is to change the API of OpenSMT and the main classes so that single Logic object can be used with multiple MainSolvers (still in sequential program only). Also, Logic and MainSolver (apart from SMTConfig) would be the only classes visible to the user.

**Comment by blishko** [comment]
There is still a problem with the method `MainSolver::readFormulaFromFile`, which not would not work.

File: src/options/SMTConfig.h [hunk 1]
```diff
@@ -319,6 +320,12 @@ struct SMTConfig
   vec<char*>    info_names;
   Map<const char*,Info*,StringHash,Equal<const char*> >   infoTable;
   Map<const char*,SMTOption*,StringHash,Equal<const char*> > optionTable;
+
+  bool usedForInitialization = false; // Some options can be changed only before this config is used for initialization of MainSolver
+  bool isPreInitializationOption(const char* o_name) {
+      return strcmp(o_name, o_produce_inter) == 0 || strcmp(o_name, o_produce_proofs) == 0;
```
Comment (aehyvari) [inline src/options/SMTConfig.h:326]: We should add:
 - o_sat_pure_lookahead
 - o_sat_lookahead_split
 - o_ghost_vars

Comment (blishko) [reply src/options/SMTConfig.h]: These options have been added.

**Comment by blishko** [comment]
It looks like there are also two regression tests in the splitting folder where the options for splitting are after the (set-logic) command. The question is, is it OK to pick the type of the solver first and other parameters can be set later (parameters such as split-num or split-format)? Or do these also need to be set before set-logic?

**Comment by aehyvari** [comment]
> It looks like there are also two regression tests in the splitting folder where the options for splitting are after the (set-logic) command.

Ok excellent, I'm fixing them!

> The question is, is it OK to pick the type of the solver first and other parameters can be set later (parameters such as split-num or split-format)? Or do these also need to be set before set-logic?

In the code this is not a problem, and I can't see how this would in the future cause a performance overhead.  It might actually be a good thing when thinking about a parallel implementation.  Let's keep it that way.

**Pull request merged**